crates/pop-common/src/sourcing/mod.rs

```diff
@@ -241,6 +271,205 @@ impl GitHub {
 			},
 		}
 	}
+
+	/// Performs any additional processing required to resolve the binary from a source.
+	///
+	/// Determines whether the binary already exists locally, using the latest version available,
+	/// and whether there are any newer versions available
+	///
+	/// # Arguments
+	/// * `name` - the name of the binary.
+	/// * `version` - a specific version of the binary required.
+	/// * `cache` - the cache being used.
+	async fn resolve(self, name: &str, version: Option<&str>, cache: &Path) -> Self {
+		match self {
+			Self::ReleaseArchive {
+				owner,
+				repository,
+				tag: _,
+				tag_pattern,
+				prerelease,
+				version_comparator,
+				fallback,
+				archive,
+				contents,
+				latest: _,
+			} => {
+				// Get releases, defaulting to the specified fallback version if an error.
+				let repo = crate::GitHub::new(owner.as_str(), repository.as_str());
+				let mut releases = repo.releases(prerelease).await.unwrap_or_else(|_e| {
+					// Use any specified version or fall back to the last known version.
+					let version = version.unwrap_or(fallback.as_str());
+					vec![Release {
+						tag_name: tag_pattern.as_ref().map_or_else(
+							|| version.to_string(),
+							|pattern| pattern.resolve_tag(version),
+						),
+						name: String::default(),
+						prerelease,
+						commit: None,
+						published_at: String::default(),
+					}]
+				});
+
+				// Filter releases if a tag pattern specified
+				if let Some(pattern) = tag_pattern.as_ref() {
+					releases.retain(|r| pattern.regex.is_match(&r.tag_name));
+				}
+
+				// Select versions from release tags, used for resolving the candidate versions and
+				// local binary versioning.
+				let mut binaries: HashMap<_, _> = releases
+					.into_iter()
+					.map(|r| {
+						let version = tag_pattern
+							.as_ref()
+							.and_then(|pattern| pattern.version(&r.tag_name).map(|v| v.to_string()))
+							.unwrap_or_else(|| r.tag_name.clone());
+						(version, r.tag_name)
+					})
+					.collect();
+
+				// Resolve any specified version - i.e., the version could be provided as a concrete
+				// version or just a tag.
+				let version = version.map(|v| {
+					tag_pattern
+						.as_ref()
+						.and_then(|pattern| pattern.version(v))
+						.unwrap_or(v)
+						.to_string()
+				});
+
+				// Extract versions from any cached binaries - e.g., offline or rate-limited.
+				for file in read_dir(cache).into_iter().flatten().filter_map(|f| {
+					f.ok().and_then(|f| f.file_name().into_string().ok()).filter(|n| {
+						n.starts_with(name) &&
+							// TODO: improve hacky workaround for naming clashes
```

Comment:
How are you thinking to refactor this workaround?

Reply:
May not be worth the effort tbh, will leave open and revisit before final review.

Reply:
Refactored in 7d629ce4 to be more extensible/configurable.